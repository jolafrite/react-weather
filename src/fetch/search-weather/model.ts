export interface ISearchWeatherFetchResponse {
  message:	string;
  cod:	number
  count:	number
  list: ISearchWeather[]
}

export interface ISearchWeather {
  id: number;
  name: string;
  coord: {
    lat: number;
    lon: number;
  };
  main: {
    temp: number;
    pressure: number;
    humidity: number;
    temp_min: number;
    temp_max: number;
    sea_level: number;
    grnd_level: number;
    temp_kf: number;
  };
  dt: number;
  wind: {
    speed: number;
    deg: number;
  };
  sys: {
    country: string
  };
  rain: {
    "1h": number;
    "3h": number;
  };
  snow: {
    "1h": number;
    "3h": number;
  };
  clouds: {
    all: number;
  };
  weather: {
    id: number;
    main: string;
    description: string;
    icon: string
  }[];
}
