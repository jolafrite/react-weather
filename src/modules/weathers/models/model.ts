export interface IFetchWeather extends IWeather {}

export interface IWeather {
  city: {
    id: number;
    name: string;
    coord: {
      lon: number;
      lat: number;
    };
    country: string;
  };
  cnt: number;
  list: IWeatherList[];
};

export interface IWeatherList {
  dt: number;
  main: {
    temp: number;
    temp_min: number;
    temp_max: number;
    pressure: number;
    humidity: number;
    sea_level: number;
    grnd_level: number;
    temp_kf: number;
  };
  weather: IWeatherListWeather[];
  clouds: {
    all: number;
  };
  wind: {
    speed: number;
    deg: number;
  };
  dt_txt: string;
};

export interface IWeatherListWeather {
  id: number;
  main: string;
  description: string;
  icon: string;
}

export interface IWeatherLast5DaysForecast {
  [date: string]: IWeatherDailyForecast;
}

export interface IWeatherDailyForecast {
  temp: number;
  temp_max: number;
  temp_min: number;
  hourly: {
    [time: string]: IWeatherList;
  };
}
