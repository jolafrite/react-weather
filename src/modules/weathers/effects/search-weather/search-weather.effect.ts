import { WeatherSearchPageActions, WeatherApiActions } from '../../actions';
import { Dispatch } from 'redux';
import { searchWeatherFetch } from '../../../../fetch';

export const searchPageSearchWeatherEffect = (query: string) => async (dispatch: Dispatch) => {
  dispatch(WeatherSearchPageActions.search(query));

  return await searchWeathersEffect(
    query,
    WeatherSearchPageActions.searchSuccess,
    WeatherSearchPageActions.searchError,
    dispatch
  );
}

const searchWeathersEffect = async (query: string, onSuccess: any, onError: any, dispatch: Dispatch) => {
  try {
    const response = await searchWeatherFetch(query);

    dispatch(WeatherApiActions.loadListSuccess(response));
    dispatch(onSuccess(response));
  } catch(e) {
    dispatch(onError(e));
  }
};
