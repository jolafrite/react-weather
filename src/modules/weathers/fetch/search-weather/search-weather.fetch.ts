import Axios from "axios";
import {
  IWeatherMain,
  IWeatherWind,
  IWeatherRain,
  IWeatherSnow,
  IWeatherClouds,
  IWeatherWeather
} from "../../models";

export interface ISearchWeatherFetchResponse {
  message: string;
  cod: string;
  count: number;
  list: {
    id: number;
    name: string;
    coord: {
      lat: number;
      lon: number;
    };
    main: IWeatherMain;
    dt: number;
    wind: IWeatherWind;
    sys: {
      country: string;
    };
    rain?: IWeatherRain | null;
    snow?: IWeatherSnow | null;
    clouds: IWeatherClouds;
    weather: IWeatherWeather[];
  }[];
}

export const searchWeatherFetch = async (query: string) => {
  const { data } = await Axios.request<ISearchWeatherFetchResponse>({
    method: "get",
    url: `${process.env.WEATHER_URL}/find`,
    params: {
      q: query,
      units: "metric",
      appid: "{WEATHER_APPID}"
    }
  });

  return data;
};
