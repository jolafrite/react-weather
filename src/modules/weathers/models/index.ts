export * from './weather.fixtures';
export * from './model';
