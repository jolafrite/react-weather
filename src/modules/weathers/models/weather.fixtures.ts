import { useFixtureFactory } from "../../../common/utils/test";
import { IWeather } from "./model";
import { ISearchWeather } from "../../../fetch/search-weather/model";

export const weatherFixtures = useFixtureFactory<IWeather>({
  city: {
    id: 1851632,
    name: "Shuzenji",
    coord: { lon: 138.933334, lat: 34.966671 },
    country: "JP"
  },
  cnt: 38,
  list: [
    {
      dt: 1406106000,
      main: {
        temp: 298.77,
        temp_min: 298.77,
        temp_max: 298.774,
        pressure: 1005.93,
        sea_level: 1018.18,
        grnd_level: 1005.93,
        humidity: 87,
        temp_kf: 0.26
      },
      weather: [
        { id: 804, main: "Clouds", description: "overcast clouds", icon: "04d" }
      ],
      clouds: { all: 88 },
      wind: { speed: 5.71, deg: 229.501 },
      dt_txt: "2014-07-23 09:00:00"
    }
  ]
});

export const searchWeatherFixtures = useFixtureFactory<ISearchWeather>({
  id: 2643743,
  name: "London",
  coord: { lat: 51.5085, lon: -0.1258 },
  main: {
    temp: 284.31,
    pressure: 1010,
    humidity: 87,
    temp_min: 283.15,
    temp_max: 285.93,
    sea_level: 1,
    grnd_level: 2,
    temp_kf: 3,
  },
  dt: 1560233226,
  wind: {
    speed: 1.5,
    deg: 150
  },
  sys: {
    country: "GB"
  },
  rain: {
    "1h": 3,
    "3h": 4,
  },
  snow: {
    "1h": 1,
    "3h": 2,
  },
  clouds: { all: 90 },
  weather: [
    { id: 500, main: "Rain", description: "light rain", icon: "10d" },
    { id: 701, main: "Mist", description: "mist", icon: "50d" }
  ]
});
