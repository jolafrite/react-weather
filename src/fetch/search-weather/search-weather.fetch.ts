import Axios from "axios";
import { IWeather } from "../../modules/weathers/models";
import { ISearchWeatherFetchResponse } from "./model";
import { searchWeatherListTransformer } from "../../modules/weathers/transformers";

export const searchWeatherFetch = async (query: string): Promise<IWeather[]> => {
  //qery must be minimum 3 characters long
  if(query.length < 3) {
    return [];
  };

  const { data } = await Axios.request<ISearchWeatherFetchResponse>({
    method: 'get',
    url: `${process.env.WEATHER_URL}/find`,
    params: {
      q: query,
      units: 'metric',
      appid: '{WEATHER_APPID}',
    }
  });

  return searchWeatherListTransformer(data.list);
};
