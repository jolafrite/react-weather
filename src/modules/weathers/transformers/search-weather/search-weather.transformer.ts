import { ISearchWeather } from "../../../../fetch/search-weather/model";
import { IWeather } from "../../models";
import { utcToDateTimeString } from "../../../../common/utils/date";

export const searchWeatherListTransformer = (weathers: ISearchWeather[]): IWeather[] => {
  return weathers.map(weather => searchWeatherTransformer(weather));
}

export const searchWeatherTransformer = (weather: ISearchWeather): IWeather => {
  return {
    city: {
      id: weather.id,
      name: weather.name,
      coord: {
        ...weather.coord
      },
      country: weather.sys.country
    },
    cnt: 1,
    list: [
      {
        dt: weather.dt,
        main: {
          ...weather.main
        },
        weather: [
          ...weather.weather,
        ],
        clouds: {
          ...weather.clouds
        },
        wind: {
          ...weather.wind,
        },
        dt_txt: utcToDateTimeString(weather.dt),
      }
    ]
  };
}
