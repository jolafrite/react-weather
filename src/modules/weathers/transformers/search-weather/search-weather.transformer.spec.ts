import { searchWeatherFixtures } from "../../models";
import { searchWeatherTransformer, searchWeatherListTransformer } from "./search-weather.transformer";

describe('searchWeatherTransformer', () => {

  test('should transform a search weather result to weather', () => {
    const weather = searchWeatherFixtures()

    expect(
      searchWeatherTransformer(weather)
    ).toMatchSnapshot();
  });

  test('should transform a list of search weather result to weather', () => {
    const weathers = [
      searchWeatherFixtures(),
      searchWeatherFixtures({ id: 1111 }),
    ];

    expect(
      searchWeatherListTransformer(weathers)
    ).toMatchSnapshot();
  });

});
