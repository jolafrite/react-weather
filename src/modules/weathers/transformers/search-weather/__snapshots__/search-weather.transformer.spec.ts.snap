// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchWeatherTransformer should transform a list of search weather result to weather 1`] = `
Array [
  Object {
    "city": Object {
      "coord": Object {
        "lat": 51.5085,
        "lon": -0.1258,
      },
      "country": "GB",
      "id": 2643743,
      "name": "London",
    },
    "cnt": 1,
    "list": Array [
      Object {
        "clouds": Object {
          "all": 90,
        },
        "dt": 1560233226,
        "dt_txt": "2019-06-11 06:07:06",
        "main": Object {
          "grnd_level": 2,
          "humidity": 87,
          "pressure": 1010,
          "sea_level": 1,
          "temp": 284.31,
          "temp_kf": 3,
          "temp_max": 285.93,
          "temp_min": 283.15,
        },
        "weather": Array [
          Object {
            "description": "light rain",
            "icon": "10d",
            "id": 500,
            "main": "Rain",
          },
          Object {
            "description": "mist",
            "icon": "50d",
            "id": 701,
            "main": "Mist",
          },
        ],
        "wind": Object {
          "deg": 150,
          "speed": 1.5,
        },
      },
    ],
  },
  Object {
    "city": Object {
      "coord": Object {
        "lat": 51.5085,
        "lon": -0.1258,
      },
      "country": "GB",
      "id": 1111,
      "name": "London",
    },
    "cnt": 1,
    "list": Array [
      Object {
        "clouds": Object {
          "all": 90,
        },
        "dt": 1560233226,
        "dt_txt": "2019-06-11 06:07:06",
        "main": Object {
          "grnd_level": 2,
          "humidity": 87,
          "pressure": 1010,
          "sea_level": 1,
          "temp": 284.31,
          "temp_kf": 3,
          "temp_max": 285.93,
          "temp_min": 283.15,
        },
        "weather": Array [
          Object {
            "description": "light rain",
            "icon": "10d",
            "id": 500,
            "main": "Rain",
          },
          Object {
            "description": "mist",
            "icon": "50d",
            "id": 701,
            "main": "Mist",
          },
        ],
        "wind": Object {
          "deg": 150,
          "speed": 1.5,
        },
      },
    ],
  },
]
`;

exports[`searchWeatherTransformer should transform a search weather result to weather 1`] = `
Object {
  "city": Object {
    "coord": Object {
      "lat": 51.5085,
      "lon": -0.1258,
    },
    "country": "GB",
    "id": 2643743,
    "name": "London",
  },
  "cnt": 1,
  "list": Array [
    Object {
      "clouds": Object {
        "all": 90,
      },
      "dt": 1560233226,
      "dt_txt": "2019-06-11 06:07:06",
      "main": Object {
        "grnd_level": 2,
        "humidity": 87,
        "pressure": 1010,
        "sea_level": 1,
        "temp": 284.31,
        "temp_kf": 3,
        "temp_max": 285.93,
        "temp_min": 283.15,
      },
      "weather": Array [
        Object {
          "description": "light rain",
          "icon": "10d",
          "id": 500,
          "main": "Rain",
        },
        Object {
          "description": "mist",
          "icon": "50d",
          "id": 701,
          "main": "Mist",
        },
      ],
      "wind": Object {
        "deg": 150,
        "speed": 1.5,
      },
    },
  ],
}
`;
